### PR TITLE
rfc: api middleware

### PR DESCRIPTION
**Pull request by Chriscbr**
[Rendered version](https://github.com/winglang/wing/blob/rybickic/api-middleware-rfc/docs/contributing/999-rfcs/2024-02-27-api-middleware.md)

Related to https://github.com/winglang/wing/issues/3518

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: docs/contributing/999-rfcs/2024-02-27-api-middleware.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+Related to the above, we would like to support the SDK exporting a type `http.Handler` that is simply a type alias for `inflight (http.Request): http.Response` to reduce the amount of boilerplate in middleware code.
+
+Second, we would like to propose an `addMiddleware` method that will allow an inflight function to be automatically applied to all routes:
+
+```js
+api.addMiddleware(loggingMiddleware); // runs on all routes
+```
+
+Lastly, to support middleware being shared between Wing libraries, we would like to prioritize fixes for the two bugs mentioned in the Design section:
+- Fix issue with reassignment to variables typed as inflight functions.
```
Comment (eladb) [inline docs/contributing/999-rfcs/2024-02-27-api-middleware.md:184]: Links?

File: docs/contributing/999-rfcs/2024-02-27-api-middleware.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+bring cloud;
+bring http;
+
+let api = new cloud.Api();
+
+let authMiddleware = (handler: http.Handler): http.Handler => {
+  return inflight (req) => {
+    let token = req.headers?.tryGet("x-service-token");
+    if let token = token {
+      return handler(req);
```
Comment (eladb) [inline docs/contributing/999-rfcs/2024-02-27-api-middleware.md:199]: How do we address the need for the user handler to obtain context from the middleware?

In this example, the user handler might want to access the token.

I am wondering if we need some arbitrary context map at the request level that middleware will be able to enrich and then the consumer will be able to access either directly (`req.ctx.get("token")`) or via a helper (`MyAuthMiddleware.getToken(req)`).

**Review by eladb — APPROVED** [review]
Nice!

**Comment by ekeren** [comment]
@Chriscbr fyi

**Pull request closed without merging**